File: backend/models/budget_model.py
from .database import get_db
import uuid
from datetime import datetime, date, timedelta
from decimal import Decimal
import json
import time

class BudgetModel:
    @staticmethod
    def get_user_budgets(user_id, active_only=True):
        """Get all budgets for a user"""
        # #region agent log
        try:
            with open(r'd:\CODE\Project\FinancialApp\financial_app\.cursor\debug.log', 'a', encoding='utf-8') as f:
                f.write(json.dumps({"id":f"log_{int(time.time()*1000)}_get_budgets_start","timestamp":int(time.time()*1000),"location":"budget_model.py:8","message":"get_user_budgets called","data":{"user_id":user_id,"active_only":active_only},"sessionId":"debug-session","runId":"run1","hypothesisId":"A"}) + "\n")
        except: pass
        # #endregion
        db = get_db()
        with db.cursor() as cursor:
            if active_only:
                sql = """
                SELECT 
                    budget_id_232143,
                    user_id_232143,
                    category_id_232143,
                    amount_232143,
                    period_232143,
                    period_start_232143,
                    period_end_232143,
                    spent_amount_232143,
                    remaining_amount_232143,
                    rollover_enabled_232143,
                    alert_threshold_232143,
                    is_active_232143,
                    recommended_amount_232143,
                    recommendation_reason_232143,
                    created_at_232143,
                    updated_at_232143
                FROM budgets_232143 
                WHERE user_id_232143 = %s AND is_active_232143 = TRUE
                ORDER BY period_start_232143 DESC
                """
            else:
                sql = """
                SELECT 
                    budget_id_232143,
                    user_id_232143,
                    category_id_232143,
                    amount_232143,
                    period_232143,
                    period_start_232143,
                    period_end_232143,
                    spent_amount_232143,
                    remaining_amount_232143,
                    rollover_enabled_232143,
                    alert_threshold_232143,
                    is_active_232143,
                    recommended_amount_232143,
                    recommendation_reason_232143,
                    created_at_232143,
                    updated_at_232143
                FROM budgets_232143 
                WHERE user_id_232143 = %s
                ORDER BY period_start_232143 DESC
                """
            # #region agent log
            try:
                with open(r'd:\CODE\Project\FinancialApp\financial_app\.cursor\debug.log', 'a', encoding='utf-8') as f:
                    f.write(json.dumps({"id":f"log_{int(time.time()*1000)}_before_execute","timestamp":int(time.time()*1000),"location":"budget_model.py:58","message":"Before executing SQL query","data":{"active_only":active_only,"sql_has_true":"TRUE" in sql},"sessionId":"debug-session","runId":"run1","hypothesisId":"A"}) + "\n")
            except: pass
            # #endregion
            cursor.execute(sql, (user_id,))
            # #region agent log
            try:
                results = cursor.fetchall()
                with open(r'd:\CODE\Project\FinancialApp\financial_app\.cursor\debug.log', 'a', encoding='utf-8') as f:
                    f.write(json.dumps({"id":f"log_{int(time.time()*1000)}_after_execute","timestamp":int(time.time()*1000),"location":"budget_model.py:60","message":"After executing SQL query","data":{"result_count":len(results)},"sessionId":"debug-session","runId":"run1","hypothesisId":"A"}) + "\n")
            except: pass
            # #endregion
            return results
    
    @staticmethod
    def get_budget_by_id(budget_id, user_id):
        """Get a specific budget by ID"""
        db = get_db()
        with db.cursor() as cursor:
            sql = """
            SELECT 
                budget_id_232143,
                user_id_232143,
                category_id_232143,
                amount_232143,
                period_232143,
                period_start_232143,
                period_end_232143,
                spent_amount_232143,
                remaining_amount_232143,
                rollover_enabled_232143,
                alert_threshold_232143,
                is_active_232143,
                recommended_amount_232143,
                recommendation_reason_232143,
                created_at_232143,
                updated_at_232143
            FROM budgets_232143 
            WHERE budget_id_232143 = %s AND user_id_232143 = %s
            """
            cursor.execute(sql, (budget_id, user_id))
            return cursor.fetchone()
    
    @staticmethod
    def create_budget(budget_data):
        """Create a new budget"""
        db = get_db()
        with db.cursor() as cursor:
            budget_id = str(uuid.uuid4())
            
            # Calculate period dates if not provided
            period = budget_data['period']
            start_date = budget_data.get('period_start')
            
            if start_date is None:
                start_date = date.today()
            elif isinstance(start_date, str):
                start_date = datetime.strptime(start_date, '%Y-%m-%d').date()
            
            # Calculate end date based on period
            if period == 'daily':
                end_date = start_date
            elif period == 'weekly':
                end_date = start_date + timedelta(days=6)
            elif period == 'monthly':
                # Get last day of month
                if start_date.month == 12:
                    end_date = date(start_date.year + 1, 1, 1) - timedelta(days=1)
                else:
                    end_date = date(start_date.year, start_date.month + 1, 1) - timedelta(days=1)
            elif period == 'yearly':
                end_date = date(start_date.year, 12, 31)
            else:
                end_date = start_date + timedelta(days=30)
            
            sql = """
            INSERT INTO budgets_232143 (
                budget_id_232143,
                user_id_232143,
                category_id_232143,
                amount_232143,
                period_232143,
                period_start_232143,
                period_end_232143,
                rollover_enabled_232143,
                alert_threshold_232143,
                is_active_232143
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            cursor.execute(sql, (
                budget_id,
                budget_data['user_id'],
                budget_data.get('category_id'),
                budget_data['amount'],
                period,
                start_date,
                end_date,
                budget_data.get('rollover_enabled', False),
                budget_data.get('alert_threshold', 80),
                budget_data.get('is_active', True)
            ))
            db.commit()
            return budget_id
    
    @staticmethod
    def update_budget(budget_id, user_id, update_data):
        """Update an existing budget"""
        db = get_db()
        with db.cursor() as cursor:
            set_clauses = []
            values = []
            
            for field, value in update_data.items():
                set_clauses.append(f"{field} = %s")
                values.append(value)
            
            if not set_clauses:
                return False
            
            values.extend([budget_id, user_id])
            
            sql = f"""
            UPDATE budgets_232143
            SET {', '.join(set_clauses)}
            WHERE budget_id_232143 = %s AND user_id_232143 = %s
            """
            
            cursor.execute(sql, values)
            db.commit()
            return cursor.rowcount > 0
    
    @staticmethod
    def delete_budget(budget_id, user_id):
        """Delete a budget"""
        db = get_db()
        with db.cursor() as cursor:
            sql = """
            DELETE FROM budgets_232143
            WHERE budget_id_232143 = %s AND user_id_232143 = %s
            """
            cursor.execute(sql, (budget_id, user_id))
            db.commit()
            return cursor.rowcount > 0
    
    @staticmethod
    def get_budgets_summary(user_id):
        """Get summary statistics for user's budgets"""
        db = get_db()
        with db.cursor() as cursor:
            sql = """
            SELECT 
                COUNT(*) as total_budgets,
                SUM(amount_232143) as total_budget,
                SUM(spent_amount_232143) as total_spent,
                SUM(remaining_amount_232143) as total_remaining,
                AVG(CASE WHEN amount_232143 > 0 
                    THEN (spent_amount_232143 / amount_232143) * 100 
                    ELSE 0 END) as avg_usage_percentage
            FROM budgets_232143
            WHERE user_id_232143 = %s AND is_active_232143 = TRUE
            """
            cursor.execute(sql, (user_id,))
            return cursor.fetchone()

File: lib/models/budget_model.dart
/// Budget Model
class BudgetModel {
  final String id;
  final String userId;
  final String categoryId;
  final double amount;
  final double spent;
  final double remaining;
  final String period;
  final DateTime periodStart;
  final DateTime periodEnd;
  final bool rolloverEnabled;
  final int alertThreshold;
  final bool isActive;
  final String? recommendationReason;
  final DateTime createdAt;
  final DateTime updatedAt;

  BudgetModel({
    required this.id,
    required this.userId,
    required this.categoryId,
    required this.amount,
    this.spent = 0.0,
    this.remaining = 0.0,
    this.period = 'monthly',
    required this.periodStart,
    required this.periodEnd,
    this.rolloverEnabled = false,
    this.alertThreshold = 80,
    this.isActive = true,
    this.recommendationReason,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) : createdAt = createdAt ?? DateTime.now(),
       updatedAt = updatedAt ?? DateTime.now();

  /// From DB map (suffixed keys)
  factory BudgetModel.fromMap(Map<String, dynamic> map) {
    return BudgetModel(
      id: (map['budget_id_232143'] ?? map['budget_id'] ?? map['id'] ?? '').toString(),
      userId: (map['user_id_232143'] ?? map['user_id'] ?? '').toString(),
      categoryId: (map['category_id_232143'] ?? map['category_id'] ?? '').toString(),
      amount: (map['amount_232143'] as num?)?.toDouble() ?? (map['amount'] as num?)?.toDouble() ?? 0.0,
      spent:
          (map['spent_amount_232143'] as num?)?.toDouble() ??
          (map['spent_amount'] as num?)?.toDouble() ??
          (map['spent'] as num?)?.toDouble() ??
          0.0,
      remaining:
          (map['remaining_amount_232143'] as num?)?.toDouble() ??
          (map['remaining_amount'] as num?)?.toDouble() ??
          (map['remaining'] as num?)?.toDouble() ??
          0.0,
      period: (map['period_232143'] ?? map['period'] ?? 'monthly').toString(),
      periodStart:
          map['period_start_232143'] != null
              ? DateTime.parse(map['period_start_232143'].toString())
              : map['period_start'] != null
              ? DateTime.parse(map['period_start'].toString())
              : DateTime.now(),
      periodEnd:
          map['period_end_232143'] != null
              ? DateTime.parse(map['period_end_232143'].toString())
              : map['period_end'] != null
              ? DateTime.parse(map['period_end'].toString())
              : DateTime.now(),
      rolloverEnabled:
          (map['rollover_enabled_232143'] ?? map['rollover_enabled'] ?? 0) == 1 ||
          (map['rollover_enabled_232143'] ?? map['rollover_enabled'] ?? false) == true,
      alertThreshold:
          (map['alert_threshold_232143'] as num?)?.toInt() ?? (map['alert_threshold'] as num?)?.toInt() ?? 80,
      isActive: map['is_active_232143'] != null ? map['is_active_232143'] == 1 : (map['is_active'] as bool?) ?? true,
      recommendationReason: (map['recommendation_reason_232143'] ?? map['recommendation_reason'])?.toString(),
      createdAt:
          map['created_at_232143'] != null ? DateTime.parse(map['created_at_232143'].toString()) : DateTime.now(),
      updatedAt:
          map['updated_at_232143'] != null ? DateTime.parse(map['updated_at_232143'].toString()) : DateTime.now(),
    );
  }

  /// To DB map (suffixed keys for inserts)
  Map<String, dynamic> toMap() {
    return {
      'budget_id_232143': id,
      'user_id_232143': userId,
      'category_id_232143': categoryId,
      'amount_232143': amount,
      'spent_amount_232143': spent,
      'remaining_amount_232143': amount - spent,
      'period_232143': period,
      'period_start_232143': periodStart.toIso8601String(),
      'period_end_232143': periodEnd.toIso8601String(),
      'rollover_enabled_232143': rolloverEnabled ? 1 : 0,
      'alert_threshold_232143': alertThreshold,
      'is_active_232143': isActive ? 1 : 0,
      if (recommendationReason != null) 'recommendation_reason_232143': recommendationReason,
      'created_at_232143': createdAt.toIso8601String(),
      'updated_at_232143': updatedAt.toIso8601String(),
    };
  }

  /// Clean JSON for serialization
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'category_id': categoryId,
      'amount': amount,
      'spent': spent,
      'remaining': amount - spent,
      'period': period,
      'period_start': periodStart.toIso8601String(),
      'period_end': periodEnd.toIso8601String(),
      'rollover_enabled': rolloverEnabled,
      'alert_threshold': alertThreshold,
      'is_active': isActive,
      'recommendation_reason': recommendationReason,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  /// Computed helpers
  double get remainingAmount => amount - spent;
  double get percentageUsed => amount > 0 ? (spent / amount) * 100 : 0.0;
  bool get isOverBudget => spent > amount;

  BudgetModel copyWith({
    String? id,
    String? userId,
    String? categoryId,
    double? amount,
    double? spent,
    double? remaining,
    String? period,
    DateTime? periodStart,
    DateTime? periodEnd,
    bool? rolloverEnabled,
    int? alertThreshold,
    bool? isActive,
    String? recommendationReason,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return BudgetModel(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      categoryId: categoryId ?? this.categoryId,
      amount: amount ?? this.amount,
      spent: spent ?? this.spent,
      remaining: remaining ?? this.remaining,
      period: period ?? this.period,
      periodStart: periodStart ?? this.periodStart,
      periodEnd: periodEnd ?? this.periodEnd,
      rolloverEnabled: rolloverEnabled ?? this.rolloverEnabled,
      alertThreshold: alertThreshold ?? this.alertThreshold,
      isActive: isActive ?? this.isActive,
      recommendationReason: recommendationReason ?? this.recommendationReason,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
